### PR TITLE
Edit rich relationship

### PR DESCRIPTION
**Pull request by kiyaGu**
This is a working version (while ignoring errors like ```Warning: Encountered two children with the same key, `2`. Keys should be unique so that components maintain their identity across updates``` and the consequence of these errors when updating records)
TODO
- fix ☝️ (the error is fixed but still get error on saving and this will be fixed in the next ticket)

### CODE REVIEW

File: packages/tc-ui/src/lib/components/input-wrapper.jsx [hunk 1]
```diff
@@ -50,7 +54,7 @@ const FieldTitle = ({ label, description, expandableContent, lockedBy }) => (
 const WrappedEditComponent = props => {
 	props = { ...props, disabled: !!props.lockedBy };
 	const { Component } = props;
-	const WrapperTag = props.wrapperTag || 'label';
+	const WrapperTag = props.wrapperTag || 'div';
```
Comment (wheresrhys) [inline packages/tc-ui/src/lib/components/input-wrapper.jsx]: is this to fix the bug with the target area for the button being the whole row?

File: packages/tc-ui/src/lib/components/input-wrapper.jsx [hunk 2]
```diff
@@ -14,7 +14,11 @@ const FieldTitle = ({ label, description, expandableContent, lockedBy }) => (
 		<span className="o-forms-title__prompt description-text">
 			<span
 				dangerouslySetInnerHTML={{
-					__html: autolinker.link(description),
+					__html:
+						typeof window === 'undefined'
+							? autolinker.link(description)
```
Comment (wheresrhys) [inline packages/tc-ui/src/lib/components/input-wrapper.jsx]: Interested in why this is necessary. Strange

File: packages/tc-ui/src/lib/mappers/component-assigner.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	getPrimitiveTypes,
+} = require('@financial-times/tc-schema-sdk');
+
+module.exports = ({ customComponents = {}, customTypeMappings = {} } = {}) => ({
+	type,
+	hasMany,
+}) => {
+	// TO Do - don't require in here
+	// eslint-disable-next-line global-require
+	const primitives = require('../../primitives/server');
```
Comment (wheresrhys) [inline packages/tc-ui/src/lib/mappers/component-assigner.js]: why can this not be required at the top level?

File: packages/tc-ui/src/pages/edit/template.jsx [hunk 4]
```diff
@@ -138,6 +137,14 @@ const EditForm = props => {
...
 			</form>
 			<script src="https://cloud.tinymce.com/stable/tinymce.js" defer />
+			<script
+				type="application/json"
+				data-json="schema-data"
+				dangerouslySetInnerHTML={{
+					__html: JSON.stringify({ ...rawData.getAll() }),
+				}}
+			/>
+			<script src="https://cdnjs.cloudflare.com/ajax/libs/autolinker/3.11.1/Autolinker.min.js" />
```
Comment (wheresrhys) [inline packages/tc-ui/src/pages/edit/template.jsx]: ah, now I see. If we can, it would be good to understand why autolinker cannot be bundled by webpack successfully

File: packages/tc-ui/src/primitives/multiple-choice/server.jsx [hunk 5]
```diff
@@ -11,7 +11,7 @@ const Checkbox = ({ name, checkboxValue, disabled, checked }) => {
 				value={checkboxValue}
 				aria-label={checkboxValue}
 				id={`checkbox-${name}-${checkboxValue}`}
-				checked={checked}
+				defaultChecked={checked}
```
Comment (wheresrhys) [inline packages/tc-ui/src/primitives/multiple-choice/server.jsx]: is this some react feature?

File: packages/tc-ui/src/server.js [hunk 6]
```diff
@@ -1,6 +1,7 @@
 const primitives = require('./primitives/server');
-
-const { componentAssigner, graphqlQueryBuilder } = require('./lib/mappers');
+const componentAssigner = require('./lib/mappers/component-assigner');
```
Comment (wheresrhys) [inline packages/tc-ui/src/server.js]: nice idea splitting these

File: example-schema/relationships/CuriousChild.yaml [hunk 7]
```diff
@@ -36,8 +36,3 @@ properties:
     label: Label for some float
     description: Description for some float.
     type: Float
-  someMultipleChoice:
```
Comment (wheresrhys) [inline example-schema/relationships/CuriousChild.yaml]: why is this removed?

File: package.json [hunk 8]
```diff
@@ -2,7 +2,7 @@
   "name": "treecreeper",
   "private": true,
   "version": "0.0.0",
-  "description": "A set of tools for building adnmin interfaces on top of graph data",
```
Comment (wheresrhys) [inline package.json]: :joy:

File: packages/tc-ui/src/lib/mappers/component-assigner.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
+const {
```
Comment (wheresrhys) [inline packages/tc-ui/src/lib/mappers/component-assigner.js]: Slightly prefer to export `{ componentAssigner }` not `componentAssigner`.

Dora persuaded me that this means there is greater consistency when importing, easier refactoring if you need to add more exports to the same file, and can be easier to write tests for

File: packages/tc-ui/src/lib/mappers/get-value.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
+const sortBy = require('lodash.sortby');
+
+module.exports = (itemSchema, itemValue) => {
```
Comment (wheresrhys) [inline packages/tc-ui/src/lib/mappers/get-value.js]: similar comment to on previous file, but also add that if you use

```
const myFunc = // define the function

module.exports = { myFunc }

```

then error stack traces and debugger tools will have information about the function name, rather than anonymous function

File: packages/tc-ui/src/pages/edit/browser.js [hunk 11]
```diff
@@ -1,9 +1,17 @@
...
+const { init } = require('@financial-times/tc-schema-sdk');
 const {
 	Relationship: { withEditComponent: attachRelationshipPicker },
 	LargeText: { withEditComponent: attachDocumentEditor },
 } = require('../../primitives/browser');
 
+const initSchema = async () => {
+	const jsonNode = document.querySelector("[data-json='schema-data']");
+	const jsonRaw = jsonNode && jsonNode.textContent;
+	const jsonData = jsonRaw && jsonRaw.length > 4 ? JSON.parse(jsonRaw) : {};
```
Comment (wheresrhys) [inline packages/tc-ui/src/pages/edit/browser.js]: Why checking for `jsonRaw.length > 4` ?

File: packages/tc-ui/src/primitives/boolean/server.jsx [hunk 12]
```diff
@@ -17,7 +17,7 @@ const Checkbox = ({ name, checkboxValue, disabled, userValue }) => {
 				value={checkboxValue.toString()}
 				aria-label={label}
 				id={`radio-${name}-${label}`}
-				checked={userValue === checkboxValue ? 'true' : null}
+				defaultChecked={userValue === checkboxValue ? 'true' : null}
```
Comment (wheresrhys) [inline packages/tc-ui/src/primitives/boolean/server.jsx]: Similar to the select defaultChecked only needing us to pass in the actual user value, ios this the correct thing to do here? I'm not sure
```suggestion
				defaultChecked={userValue}
```

File: packages/tc-ui/src/primitives/relationship/lib/relationship-picker.jsx [hunk 13]
```diff
@@ -78,7 +79,7 @@ class RelationshipPicker extends React.Component {
 		// this is needed to prevent the event propagating up and then
 		// immediately clicking another button. VERY odd behaviour, and
 		// don't fully understand why, but this is the fix
-		event.preventDefault();
+		event.stopPropagation();
```
Comment (wheresrhys) [inline packages/tc-ui/src/primitives/relationship/lib/relationship-picker.jsx]: ah - was this caused by the `label` thing too?

File: packages/tc-ui/src/primitives/text/server.jsx [hunk 14]
```diff
@@ -1,19 +1,23 @@
...
-			className="o-forms__text"
-			type="text"
-			value={value || null}
-			required={required ? 'required' : null}
-			disabled={disabled}
-		/>
-	</span>
-);
+const EditText = ({ propertyName, value, required, lockedBy, disabled }) => {
+	const name =
```
Comment (wheresrhys) [inline packages/tc-ui/src/primitives/text/server.jsx]: I guess you don't need this any more?

File: webpack.config.js [hunk 15]
```diff
@@ -38,6 +38,10 @@ module.exports = {
 			encoding: 'utf8',
 		}),
 	],
+	node: {
+		fs: 'empty',
```
Comment (wheresrhys) [inline webpack.config.js]: can you leave a comment explaining why this needs to be here?

File: packages/tc-ui/src/primitives/relationship/lib/relationship-picker.jsx (hunk 13 above)
Comment (kiyaGu) [reply packages/tc-ui/src/primitives/relationship/lib/relationship-picker.jsx]: not sure, I this is a more appropriate than preventDefault for this scenario

File: packages/tc-ui/src/primitives/boolean/server.jsx (hunk 12 above)
Comment (kiyaGu) [reply packages/tc-ui/src/primitives/boolean/server.jsx]: I think this will not work - if you pick 'No' - when you go back to edit button it will not select it as `userValue = false` and if you picked 'Yes` in the edit page it will pick 'No' as your selection as it is the second (both Yes and No have value of defaultSelected = true)

File: example-schema/relationships/CuriousChild.yaml (hunk 7 above)
Comment (kiyaGu) [reply example-schema/relationships/CuriousChild.yaml]: it is defined in line 26

File: webpack.config.js (hunk 15 above)
Comment (kiyaGu) [reply webpack.config.js]: I added that without doing much investigation but it will suppress the following error `Module not found: Error: Can't resolve 'fs'` while I still using `css-loader@3.2.1`
Screenshot 2020-02-14 at 17 46 43
shall I create a ticket to investigate that?

File: packages/tc-ui/src/pages/edit/browser.js (hunk 11 above)
Comment (kiyaGu) [reply packages/tc-ui/src/pages/edit/browser.js]: no particular reason other than guarding 'null' string

File: webpack.config.js (hunk 15 above)
Comment (kiyaGu) [reply webpack.config.js]: this is resolved now with the refactoring done

File: example-schema/relationships/CuriousChild.yaml (hunk 7 above)
Comment (wheresrhys) [reply example-schema/relationships/CuriousChild.yaml]: oh yes, I remember now

File: packages/tc-schema-sdk/server.js [hunk 16]
```diff
@@ -1,7 +1,6 @@
 const { SDK } = require('./sdk');
 const readYaml = require('./lib/read-yaml');
 
-module.exports = new SDK({ init: false });
+module.exports = new SDK({ init: false, readYaml });
 
 module.exports.SDK = SDK;
```
Comment (wheresrhys) [inline packages/tc-schema-sdk/server.js]: if you continue to export readYaml from this file...

File: packages/tc-schema-validator/tests/enums.js [hunk 17]
```diff
@@ -1,7 +1,8 @@
 /* global it, describe, expect */
 const { SDK } = require('@financial-times/tc-schema-sdk');
```
Comment (wheresrhys) [inline packages/tc-schema-validator/tests/enums.js]: then you can add `, readYaml` to this first import

File: packages/tc-schema-validator/tests/enums.js [hunk 18]
```diff
@@ -1,7 +1,8 @@
...
+const readYaml = require('@financial-times/tc-schema-sdk/lib/read-yaml');
```
Comment (wheresrhys) [inline packages/tc-schema-validator/tests/enums.js]: and remove this import.

The reason this is a good thing to do is because unless a library explicitly documents additional files are meant for requiring, it's safer/less likely to need refactoring to just require the package by name, rather than files internal to the package.

File: packages/tc-ui/src/primitives/boolean/server.jsx (hunk 12 above)
Comment (wheresrhys) [reply packages/tc-ui/src/primitives/boolean/server.jsx]: ok, I think I follow. Do we have cypress tests for his scenario?

File: packages/tc-ui/src/primitives/relationship/lib/relationship-picker.jsx (hunk 13 above)
Comment (wheresrhys) [reply packages/tc-ui/src/primitives/relationship/lib/relationship-picker.jsx]: Have you manually tested that the bad behaviour described in the comment doesn't happen after this change? If so, then :+1:

File: packages/tc-ui/src/primitives/relationship/lib/relationship.jsx [hunk 19]
```diff
@@ -1,25 +1,96 @@
...
+								disabled ? 'disabled' : ''
+							}`}
+							onClick={this.showRichRelationshipEditor}
+							data-index={`annotate-${index}`}
+						>
+							Annotate
+						</button>
+					</span>
+					{isMounted && annotate ? (
+						<RichRelationships key={index} {...this.props} />
```
Comment (wheresrhys) [inline packages/tc-ui/src/primitives/relationship/lib/relationship.jsx]: Small niggle - can this component be renamed to RichRelationship or RelationshipProperties or something. While the implementation was in flux I wasn't sure what the content of the component would end up being, but now (if I understand correctly) it is a wrapper around the properties of a single relationship

File: packages/tc-schema-validator/tests/enums.js (hunk 18 above)
Comment (kiyaGu) [reply packages/tc-schema-validator/tests/enums.js]: thanks

File: packages/tc-ui/src/primitives/relationship/lib/relationship.jsx (hunk 19 above)
Comment (kiyaGu) [reply packages/tc-ui/src/primitives/relationship/lib/relationship.jsx]: `ok, I think I follow. Do we have cypress tests for his scenario?`
Yes =>
https://github.com/Financial-Times/treecreeper/blob/6352deb1ef8352ca81cc55b447554fcf27cd3dc5/packages/tc-ui/src/pages/edit/__tests__/e2e-edit-record.cyp.js#L110

`Have you manually tested that the bad behaviour described in the comment doesn't happen after this change?`
I did and didn't see that behaviour happening ... it might be, in addition, because we have fixed the issue we had with Remove button responding to other clicks inside of `<label>` region

**Comment by kiyaGu** [comment]
Thanks for reviewing thoroughly 👍

File: packages/tc-schema-sdk/server.js [hunk 20]
```diff
@@ -1,7 +1,7 @@
 const { SDK } = require('./sdk');
 const readYaml = require('./lib/read-yaml');
 
-module.exports = new SDK({ init: false });
+module.exports = new SDK({ init: false, readYaml });
```
Comment (wheresrhys) [inline packages/tc-schema-sdk/server.js]: I just had an idea. You can use `SDK.prototype.readYaml = readYaml;`, then you never have to pass it in explicitly here or in the tests as on any instance of the SDK `this.readYaml` will be inherited from the prototype

Comment (kiyaGu) [reply packages/tc-schema-sdk/server.js]: giphy
